practice/completer/site.js

```diff
@@ -0,0 +1,56 @@
+'use strict';
+
+/**
+ *  The Completer Constructor holds valid words that can be called through
+ * the prototype method complete, to find valid words that start with given
+ * prefix.
+ * There is also a weighting system, where everytime selectCompletion is called
+ * the selected string has weight added to it. The complete function returns
+ * an array ordered in more weighted words first.
+ */
+function Completer() {
+  this.validCompletionToWeight = {};
+}
+
+var completerProto = {
+  addCompletion: function(str) {
+    this.validCompletionToWeight[str] = 0;
+  },
+  removeCompletion: function(str) {
+    delete this.validCompletionToWeight[str];
+  },
+  complete: function(prefix) {
+    var potentialWords = _.filter(
+      _.keys(this.validCompletionToWeight), function(o) {
+        return new RegExp('^' + prefix).test(o);
+      });
```

Comment:
This function creates an entirely new RegExp for each valid completion, even though they're all identical. This code will be much faster if you make a RegExp once at the top and then use it in the anonymous function.

More advanced, but if you have a function that basically just returns the call of a function on a single value, that's a no-op wrapper.

```js
_.filter(input, prefixRE.test);
```

will also work.